README.md
# NAME

WebService::MinFraud - API for MaxMind's minFraud Score, Insights, and Factors services

# VERSION

version 1.009001

# SYNOPSIS

    use 5.010;

    use WebService::MinFraud::Client;

    # The Client object can be re-used across several requests.
    # Your MaxMind account_id and license_key are available at
    # https://www.maxmind.com/en/my_license_key
    my $client = WebService::MinFraud::Client->new(
        account_id  => 42,
        license_key => 'abcdef123456',
    );

    # For the fraud services, the request HashRef must contain a 'device' key, with a value that is a
    # HashRef containing an 'ip_address' key with a valid IPv4 or IPv6 address.
    # All other keys/values are optional; see other modules in minFraud Perl API
    # distribution for details.

    my $request = { device => { ip_address => '24.24.24.24' } };

    # Use the 'score', 'insights', or 'factors' client methods, depending on
    # the minFraud web service you are using.

    my $score = $client->score( $request );
    say $score->risk_score;

    my $insights = $client->insights( $request );
    say $insights->shipping_address->is_high_risk;

    my $factors = $client->factors( $request );
    say $factors->subscores->ip_tenure;

    # For the chargeback service, the request HashRef must contain an 'ip_address' key
    # with a valid IPv4 or IPv6 address.
    # All other keys/values are optional; see other modules in minFraud Perl API
    # distribution for details.

    $request = { ip_address => '24.24.24.24' };

    # Use the 'chargeback' method. The chargeback api does not return
    # any content from the server.

    my $chargeback = $client->chargeback( $request );
    if ($chargeback->isa('WebService::MinFraud::Model::Chargeback')) {
      say 'Successfully submitted chargeback';
    }

# DESCRIPTION

This distribution provides an API for the
[MaxMind minFraud Score, Insights, and Factors web services](https://dev.maxmind.com/minfraud/)
and the [MaxMind minFraud Chargeback web service](https://dev/maxmind.com/minfraud/chargeback/).

See [WebService::MinFraud::Client](https://metacpan.org/pod/WebService::MinFraud::Client) for details on using the web service client
API.

# INSTALLATION

The minFraud Perl API and its dependencies can be installed with
[cpanm](https://metacpan.org/pod/App::cpanminus). `cpanm` itself has no
dependencies.

    cpanm WebService::MinFraud

# VERSIONING POLICY

The minFraud Perl API uses [Semantic Versioning](https://semver.org/).

# PERL VERSION SUPPORT

The minimum required Perl version for the minFraud Perl API is 5.10.0.

The data returned from the minFraud web services includes Unicode characters
in several locales. This may expose bugs in earlier versions of Perl. If
Unicode is important to your work, we recommend that you use the most recent
version of Perl available.

# SUPPORT

This module is deprecated and will only receive fixes for major bugs and
security vulnerabilities. New features and functionality will not be added.

Please report all issues with this distribution using the GitHub issue tracker
at [https://github.com/maxmind/minfraud-api-perl/issues](https://github.com/maxmind/minfraud-api-perl/issues).

If you are having an issue with a MaxMind service that is not specific to the
client API please visit [https://www.maxmind.com/en/support](https://www.maxmind.com/en/support) for details.

Bugs may be submitted through [https://github.com/maxmind/minfraud-api-perl/issues](https://github.com/maxmind/minfraud-api-perl/issues).

# AUTHOR

Mateu Hunter <mhunter@maxmind.com>

# CONTRIBUTORS

- Andy Jack <ajack@maxmind.com>
- Christopher Bothwell <christopher.bothwell@endurance.com>
- Dave Rolsky <drolsky@maxmind.com>
- Florian Ragwitz <rafl@debian.org>
- Greg Oschwald <goschwald@maxmind.com>
- Mark Fowler <mark@twoshortplanks.com>
- Narsimham Chelluri <nchelluri@maxmind.com>
- Olaf Alders <oalders@maxmind.com>
- Patrick Cronin <pcronin@maxmind.com>
- Ruben Navarro <rnavarro@maxmind.com>
- William Storey <wstorey@maxmind.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 - 2020 by MaxMind, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

1.010000 2020-05-04

- You may now use a license key object that stringifies to the actual license
  key.


1.009001 2019-06-18

- Added deprecation notice
- Added the following new values to the /payment/processor validation:
  - datacash
  - gocardless
  - paylike
  - smartdebit
  - synapsefi


1.009000 2018-09-10

- Added the following new values to the /payment/processor validation:
  - payeezy
  - payment_express
- Updated GeoIP2 dependency to GeoIP2 2.006001. This adds support for license
  keys longer than 12 characters.


1.008000 2018-05-14

[ENHANCEMENTS]

- Added a new model for WebService::MinFraud::Chargebacks
- Updated WebService::MinFraud::Client to support API calls to
   chargebacks
  - Added chargeback subroutine
- Updated How WebService::MinFraud::Validator works
  - Validation for individual API end points moved to
    WebService::MinFraud::Validator::*
  - Added validators for existing fraud services
  - Added validation for chargebacks
  - Validator now checks for the endpoint it needs to validate
  - Validator interface preserved

[ BUGFIXES ]

- Corrected autovivification bug in WebService::MinFraud::Client->_fix_booleans
- Fixed test exception in xt/author-basic.t


1.007000 2018-04-11

- Renamed the WebService::MinFraud::Client constructor argument user_id
  to account_id. user_id still works but should be considered deprecated.
- Added the following new values to the /payment/processor validation:
  - ccavenue
  - ct_payments
  - dalenys
  - oney
  - posconnect
- Added the following new value to the /event/type validation:
  payout_change
- Added the /device/local_time output.
- Added the /credit_card/is_virtual output.


1.006000 2018-01-18

- Updated GeoIP2 dependency to GeoIP2 2.005000. This adds the
  is_in_european_union() attribute to GeoIP2::Record::Country and
  GeoIP2::Record::RepresentedCountry. This attribute is true if the
  country is a member state of the European Union.
- Added the following new values to the /payment/processor validation:
  - cybersource
  - transact_pro
  - wirecard


1.005001 2017-10-30

- Updated GeoIP2 dependency to add support for GeoIP2 Precision Insights
  anonymizer fields.
- Added the following new values to the /payment/processor validation:
  - bpoint
  - checkout_com
  - emerchantpay
  - heartland
  - payway


1.005000 2017-07-07

[ENHANCEMENTS]

- Added support for custom inputs. You may set up custom inputs from your
  account portal.
- Added the following new values to the /payment/processor validation:
  - american_express_payment_gateway
  - bluesnap
  - commdoo
  - curopayments
  - ebs
  - exact
  - hipay
  - lemon_way
  - oceanpayment
  - paymentwall
  - payza
  - securetrading
  - solidtrust_pay
  - vantiv
  - vericheck
  - vpos
- Added /device/session_id, /device/session_age inputs and /email/first_seen
  output.


1.004000 2017-01-04

[ENHANCEMENTS]

- This module now uses namespace::autoclean to remove imported functions from
  the namespaces of modules.

[ BUGFIXES ]

- Updated API to work with the latest GeoIP2 release, which uses
  namespace::autoclean to remove imported functions from the namespace.
  Previously, this module relied on has() being in the namespace of the model
  classes.

[OTHER]

- Updated the docs for the WebService::MinFraud::Record::BillingAddress and
  WebService::MinFraud::Record::ShippingAddress classes' is_postal_in_city
  method to no longer say that it is US-only


1.003000 2016-11-21

[ENHANCEMENTS]

- The disposition was added to the minFraud response models. This is used to
  return the disposition of the transaction as set by the custom rules for the
  account.


1.002000 2016-11-11

[ENHANCEMENTS]

- Added the new input /credit_card/token.


1.001000 2016-10-10

[ENHANCEMENTS]

- Added the following new values to the /event/type validation:
  password_reset and email_change.


1.000001 2016-09-15

[ OTHER ]

- Remove outdated beta notice.


1.000000 2016-09-15

[ENHANCEMENTS]

- Added the following new values to the /payment/processor validation:
  concept_payments, ecomm365, orangepay, and pacnet_services.


0.004000 2016-06-08

[BREAKING CHANGE]

- credits_remaining has been removed from the web service models and has been
  replaced by queries_remaining.

[ENHANCEMENTS]

- Added queries_remaining and funds_remaining. Note that funds_remaining
  will not be returned by the web service until our new  credit system is in
  place.
- Added device confidence and last_seen to minFraud Insights


0.003000 2016-05-23

[ENHANCEMENTS]

- Added support for the minFraud Factors.
- Added IP address risk to the minFraud Score model.
- Added the following new values to the /payment/processor validation:
  ccnow, dalpay, epay (replaces epayeu), payplus, pinpayments, quickpay, and
  verepay.

[ BUGFIXES ]

- Setting the locales on the client now sets the locales in the model objects
  as documented
- An WebService::MinFraud::Error::IPAddressNotFound is no longer erroneously
  thrown if a 404 is received.
- Some bugs in the HTTP error status handling were fixed.
- When validating a request body, this module no longer validates the TLD on
  domains to prevent false errors due to new gTLDs.

[ OTHER ]

- Added tests for WebService::MinFraud::Client that uses mocked user agent.


0.002000 2016-01-21

- The warning "input" has been restructured to be "input_pointer"
- brand and type added to credit_card
- device id added
- email is_free and is_high_risk added


0.001003 2015-08-03

- Remove trivial hash values from the request for stricter validation


0.001002 2015-06-30

- Improve POD


0.001001 2015-06-29

- First release


lib/WebService/MinFraud/Client.pm
package WebService::MinFraud::Client;

use 5.010;
use Moo 1.004005;
use namespace::autoclean;

our $VERSION = '1.010001';

use HTTP::Headers ();
use HTTP::Request ();
use JSON::MaybeXS;
use LWP::UserAgent;
use Scalar::Util qw( blessed );
use Sub::Quote qw( quote_sub );
use Try::Tiny qw( catch try );
use Types::Standard qw( Defined InstanceOf );
use URI ();
use WebService::MinFraud::Error::Generic;
use WebService::MinFraud::Error::HTTP;
use WebService::MinFraud::Error::WebService;
use WebService::MinFraud::Model::Factors;
use WebService::MinFraud::Model::Insights;
use WebService::MinFraud::Model::Score;
use WebService::MinFraud::Model::Chargeback;
use WebService::MinFraud::Types qw(
    JSONObject
    MaxMindID
    Str
    URIObject
    UserAgentObject
);
use WebService::MinFraud::Validator;

with 'WebService::MinFraud::Role::HasLocales';

has account_id => (
    is       => 'ro',
    isa      => MaxMindID,
    required => 1,
);
*user_id = \&account_id;    # for backwards-compatibility

has _base_uri => (
    is      => 'lazy',
    isa     => URIObject,
    builder => sub {
        my $self = shift;
        URI->new( $self->uri_scheme . '://' . $self->host . '/minfraud' );
    },
);
has host => (
    is      => 'ro',
    isa     => Str,
    default => q{minfraud.maxmind.com},
);
has _json => (
    is       => 'ro',
    isa      => JSONObject,
    init_arg => undef,
    default  => quote_sub(q{ JSON::MaybeXS->new->utf8 }),
);
has license_key => (
    is       => 'ro',
    isa      => Defined,
    required => 1,
);

has timeout => (
    is      => 'ro',
    isa     => Str,
    default => q{},
);

has ua => (
    is      => 'lazy',
    isa     => UserAgentObject,
    builder => sub { LWP::UserAgent->new },
);

has uri_scheme => (
    is      => 'ro',
    isa     => Str,
    default => q{https},
);

has _validator => (
    is      => 'lazy',
    isa     => InstanceOf ['WebService::MinFraud::Validator'],
    builder => sub { WebService::MinFraud::Validator->new },
    handles => { _remove_trivial_hash_values => '_delete' },
);

around BUILDARGS => sub {
    my $orig = shift;

    my $args = $orig->(@_);

    $args->{account_id} = delete $args->{user_id} if exists $args->{user_id};

    return $args;
};

sub BUILD {
    my $self = shift;

    ## no critic (RequireBlockTermination)
    my $self_version = try { 'v' . $self->VERSION() } || 'v?';

    my $ua         = $self->ua();
    my $ua_version = try { 'v' . $ua->VERSION() } || 'v?';
    ## use critic

    my $agent
        = blessed($self)
        . " $self_version" . ' ('
        . blessed($ua) . q{ }
        . $ua_version . q{ / }
        . "Perl $^V)";

    $ua->agent($agent);
}

sub factors {
    my $self = shift;

    return $self->_response_for(
        'v2.0',
        'factors',
        'WebService::MinFraud::Model::Factors', @_,
    );
}

sub insights {
    my $self = shift;

    return $self->_response_for(
        'v2.0',
        'insights',
        'WebService::MinFraud::Model::Insights', @_,
    );
}

sub score {
    my $self = shift;

    return $self->_response_for(
        'v2.0',
        'score',
        'WebService::MinFraud::Model::Score', @_,
    );
}

sub chargeback {
    my $self = shift;

    return $self->_response_for(
        undef,
        'chargeback',
        'WebService::MinFraud::Model::Chargeback', @_,
    );
}

sub _response_for {
    my ( $self, $version, $path, $model_class, $content ) = @_;

    $content = $self->_remove_trivial_hash_values($content);

    $self->_fix_booleans($content);

    my $uri           = $self->_base_uri->clone;
    my @path_segments = ( $uri->path_segments, );
    push @path_segments, $version if $version;
    push @path_segments, $path;
    $uri->path_segments(@path_segments);

    $self->_validator->validate_request( $content, $path );
    my $request = HTTP::Request->new(
        'POST', $uri,
        HTTP::Headers->new(
            Accept         => 'application/json',
            'Content-Type' => 'application/json'
        ),
        $self->_json->encode($content)
    );

    $request->authorization_basic( $self->account_id, $self->license_key );

    my $response = $self->ua->request($request);

    if ( $response->code == 200 ) {
        my $body = $self->_handle_success( $response, $uri );
        return $model_class->new( %{$body}, locales => $self->locales, );
    }
    elsif ( $response->code == 204 ) {
        return $model_class->new;
    }
    else {
        # all other error codes throw an exception
        $self->_handle_error_status( $response, $uri );
    }
}

{
    my @Booleans = (
        [ 'order',   'is_gift' ],
        [ 'order',   'has_gift_message' ],
        [ 'payment', 'was_authorized' ],
    );

    sub _fix_booleans {
        my $self    = shift;
        my $content = shift;

        return unless $content;

        for my $boolean (@Booleans) {
            my ( $object, $key ) = @{$boolean};
            if (   !exists $content->{$object}
                || !exists $content->{$object}{$key} ) {
                next;
            }

            $content->{$object}{$key}
                = $content->{$object}{$key}
                ? JSON()->true
                : JSON()->false;
        }
    }
}

sub _handle_success {
    my $self     = shift;
    my $response = shift;
    my $uri      = shift;

    my $body;
    try {
        $body = $self->_json->decode( $response->decoded_content );
    }
    catch {
        WebService::MinFraud::Error::Generic->throw(
            message =>
                "Received a 200 response for $uri but could not decode the response as JSON: $_",
        );
    };

    return $body;
}

sub _handle_error_status {
    my $self     = shift;
    my $response = shift;
    my $uri      = shift;

    my $status = $response->code;

    if ( $status =~ /^4/ ) {
        $self->_handle_4xx_status( $response, $status, $uri );
    }
    elsif ( $status =~ /^5/ ) {
        $self->_handle_5xx_status( $status, $uri );
    }
    else {
        $self->_handle_non_200_status( $status, $uri );
    }
}

sub _handle_4xx_status {
    my $self     = shift;
    my $response = shift;
    my $status   = shift;
    my $uri      = shift;

    my $content = $response->decoded_content;

    my $has_body = defined $content && length $content;
    my $body     = try {
        $has_body
            && $response->content_type =~ /json/
            && $self->_json->decode($content)
    };

    if ($body) {
        if ( $body->{code} || $body->{error} ) {
            WebService::MinFraud::Error::WebService->throw(
                message => delete $body->{error},
                %{$body},
                http_status => $status,
                uri         => $uri,
            );
        }
        else {
            WebService::MinFraud::Error::Generic->throw( message =>
                    'Response contains JSON but it does not specify code or error keys'
            );
        }
    }
    else {
        WebService::MinFraud::Error::HTTP->throw(
            message => $has_body
            ? "Received a $status error for $uri with the following body: $content"
            : "Received a $status error for $uri with no body",
            http_status => $status,
            uri         => $uri,
        );
    }
}

sub _handle_5xx_status {
    my $self   = shift;
    my $status = shift;
    my $uri    = shift;

    WebService::MinFraud::Error::HTTP->throw(
        message     => "Received a server error ($status) for $uri",
        http_status => $status,
        uri         => $uri,
    );
}

sub _handle_non_200_status {
    my $self   = shift;
    my $status = shift;
    my $uri    = shift;

    WebService::MinFraud::Error::HTTP->throw(
        message =>
            "Received an unexpected HTTP status ($status) for $uri that is neither 2xx, 4xx nor 5xx",
        http_status => $status,
        uri         => $uri,
    );
}

1;

# ABSTRACT: Perl API for MaxMind's minFraud Score and Insights web services

__END__

=head1 SYNOPSIS

  use 5.010;

  use WebService::MinFraud::Client;

  # The Client object can be re-used across several requests.
  # Your MaxMind account_id and license_key are available at
  # https://www.maxmind.com/en/my_license_key
  my $client = WebService::MinFraud::Client->new(
      account_id  => 42,
      license_key => 'abcdef123456',
  );

  # Request HashRef must contain a 'device' key, with a value that is a
  # HashRef containing an 'ip_address' key with a valid IPv4 or IPv6 address.
  # All other keys/values are optional; see other modules in minFraud Perl API
  # distribution for details.

  my $request = { device => { ip_address => '24.24.24.24' } };

  # Use the 'score', 'insights', or 'factors' client methods, depending on
  # the minFraud web service you are using.

  my $score = $client->score( $request );
  say $score->risk_score;

  my $insights = $client->insights( $request );
  say $insights->shipping_address->is_high_risk;

  my $factors = $client->factors( $request );
  say $factors->subscores->ip_tenure;


  # Request HashRef must contain an 'ip_address' key containing  a valid
  # IPv4 or IPv6 address. All other keys/values are optional; see other modules
  # in minFraud Perl API distribution for details.

  $request = { ip_address => '24.24.24.24' };

  # Use the chargeback client method to submit an IP address back to Maxmind.
  # The chargeback api does not return any content from the server.

  my $chargeback = $client->chargeback( $request );
  if ($chargeback->isa('WebService::MinFraud::Model::Chargeback')) {
    say 'Successfully submitted chargeback';
  }

=head1 DESCRIPTION

This class provides a client API for the MaxMind minFraud Score, Insights
Factors web services, and the Chargeback web service. The B<Insights>
service returns more data about a transaction than the B<Score> service.
See the L<API documentation|https://dev.maxmind.com/minfraud/>
for more details.

Each web service is represented by a different model class, and
these model classes in turn contain multiple Record classes. The Record
classes have attributes which contain data about the transaction or IP address.

If the web service does not return a particular piece of data for a
transaction or IP address, the associated attribute is not populated.

The web service may not return any information for an entire record, in which
case all of the attributes for that record class will be empty.

=head1 TRANSPORT SECURITY

Requests to the minFraud web service are made over an HTTPS connection.

=head1 USAGE

The basic API for this class is the same for all of the web services. First you
create a web service object with your MaxMind C<account_id> and C<license_key>,
then you call the method corresponding to the specific web service, passing it
the transaction you want analyzed.

If the request succeeds, the method call will return a model class for the web
service you called. This model in turn contains multiple record classes, each of
which represents part of the data returned by the web service.

If the request fails, the client class throws an exception.

=head1 CONSTRUCTOR

This class has a single constructor method:

=head2 WebService::MinFraud::Client->new

This method creates a new client object. It accepts the following arguments:

=over 4

=item * account_id

Your MaxMind User ID. Go to L<https://www.maxmind.com/en/my_license_key> to see
your MaxMind User ID and license key.

This argument is required.

=item * license_key

Your MaxMind license key.

This argument is required.

=item * locales

This is an array reference where each value is a string indicating a locale.
This argument will be passed on to record classes to use when their C<name>
methods are called.

The order of the locales is significant. When a record class has multiple
names (country, city, etc.), its C<name> method will look at each element of
this array reference and return the first locale for which it has a name.

Note that the only locale which is always present in the minFraud data is
C<en>. If you do not include this locale, the C<name> method may return
C<undef> even when the record in question has an English name.

Currently, the valid list of locale codes is:

=over 8

=item * de - German

=item * en - English

English names may still include accented characters if that is the accepted
spelling in English. In other words, English does not mean ASCII.

=item * es - Spanish

=item * fr - French

=item * ja - Japanese

=item * pt-BR - Brazilian Portuguese

=item * ru - Russian

=item * zh-CN - Simplified Chinese

=back

Passing any other locale code will result in an error.

The default value for this argument is C<['en']>.

=item * host

The hostname of the minFraud web service used when making requests. This
defaults to C<minfraud.maxmind.com>. In most cases, you do not need to set this
explicitly.

=item * ua

This argument allows you to set your own L<LWP::UserAgent> object. This is
useful if you have to override the default object (C<< LWP::UserAgent->new() >>)
to set http proxy parameters, for example.

This attribute can be any object which supports C<agent> and C<request>
methods:

=over 8

=item * request

The C<request> method will be called with an L<HTTP::Request> object as its only
argument. This method must return an L<HTTP::Response> object.

=item * agent

The C<agent> method will be called with a User-Agent string, constructed as
described below.

=back

=back

=head1 REQUEST

The request methods are passed a HashRef as the only argument. See the L</SYNOPSIS> and L<WebService::MinFraud::Example> for detailed usage examples. Some important notes regarding values passed to the minFraud web service via the Perl API are described below.

=head2 device => ip_address or ip_address

This must be a valid IPv4 or IPv6 address in presentation format, i.e.,
dotted-quad notation or the IPv6 hexadecimal-colon notation.

=head1 REQUEST METHODS

All of the fraud service request methods require a device ip_address. See the
L<API documentation for fraud services|https://dev.maxmind.com/minfraud/>
for details on all the values that can be part of the request. Portions of the
request hash with undefined and empty string values are automatically removed
from the request.

The chargeback request method requires an ip_address. See the
L<API documentation for chargeback|https:://dev.maxmind.com/minfraud/chargeback/>
for details on all the values that can be part of the request.

=head2 score

This method calls the minFraud Score web service. It returns a
L<WebService::MinFraud::Model::Score> object.

=head2 insights

This method calls the minFraud Insights web service. It returns a
L<WebService::MinFraud::Model::Insights> object.

=head2 factors

This method calls the minFraud Factors web service. It returns a
L<WebService::MinFraud::Model::Factors> object.

=head2 chargeback

This method calls the minFraud Chargeback web service. It returns a
L<WebService::MinFraud::Model::Chargeback> object.

=head1 User-Agent HEADER

Requests by the minFraud Perl API will have a User-Agent header containing the
package name and version of this module (or a subclass if you use one), the
package name and version of the user agent object, and the version of Perl.

This header is set in order to help us support individual users, as well to determine
support policies for dependencies and Perl itself.

=head1 EXCEPTIONS

For details on the possible errors returned by the web service itself, please
refer to the
L<API documentation|https://dev.maxmind.com/minfraud/>.

Prior to making the request to the web service, the request HashRef is passed
to L<WebService::MinFraud::Validator> for checks. If the request fails
validation an exception is thrown, containing a string describing all of the
validation errors.

If the web service returns an explicit error document, this is thrown as a
L<WebService::MinFraud::Error::WebService> exception object. If some other
sort of error occurs, this is thrown as a L<WebService::MinFraud::Error::HTTP>
object. The difference is that the web service error includes an error message
and error code delivered by the web service. The latter is thrown when an
unanticipated error occurs, such as the web service returning a 500 status or an
invalid error document.

If the web service returns any status code besides 200, 4xx, or 5xx, this also
becomes a L<WebService::MinFraud::Error::HTTP> object.

Finally, if the web service returns a 200 but the body is invalid, the client
throws a L<WebService::MinFraud::Error::Generic> object.

All of these error classes have a C<< message >> method and
overload stringification to show that message. This means that if you don't
explicitly catch errors they will ultimately be sent to C<STDERR> with some
sort of (hopefully) useful error message.

=head1 WHAT DATA IS RETURNED?

Please see the
L<API documentation|https://dev.maxmind.com/minfraud/>
for details on what data each web service may return.

Every record class attribute has a corresponding predicate method so that you
can check to see if the attribute is set.

  my $insights = $client->insights( $request );
  my $issuer   = $insights->issuer;
  # phone_number attribute, with has_phone_number predicate method
  if ( $issuer->has_phone_number ) {
      say "issuer phone number: " . $issuer->phone_number;
  }
  else {
      say "no phone number found for issuer";
  }
